### PR TITLE
fix(core): 43 correctness bugs found by a systematic core scan

### PR DESCRIPTION
**Pull request by RaananW**
## Summary

A systematic correctness scan of `@babylonjs/core`, fixing **43 verified bugs**. Every finding was traced through the actual source — inputs walked by hand, mirror/sibling implementations compared, and consumers checked — before being accepted. Speculative, stylistic, and behaviourally-ambiguous candidates were rejected.

No public API shapes change. `tsc`, `eslint`, `prettier` and all **3065** unit tests pass.

> All bugs were re-confirmed to still be present on latest `master` (v9.20.0, merged into this branch) before fixing.

## Recurring bug classes

Most findings fall into a few repeating patterns, which is what made them findable:

- **Mirror asymmetry** — one of a pair of near-identical functions was updated and the other wasn't (`_getFirstNonNullVector`/`_getLastNonNullVector`, `_hookArrayForIncludedOnly`/`_hookArrayForExcluded`, pointer-selection vs near-interaction, Anaglyph vs Stereoscopic cameras, Min vs Max branches).
- **Copy-paste with an incomplete edit** — `Math.min` left behind in a `Max` branch, `.x/.y/.z/.w` left behind on a `Color4`, a duplicated registration key.
- **Nullable-number truthiness** — `if (x)` / `x || default` where `0` is a legal, documented value.
- **Add/remove target mismatch** — a listener or observer registered on one object and removed from another.

## Maths

| File | Bug |
|---|---|
| `Maths/math.vector.pure.ts` | `Vector2.normalizeToRef` missing `return` and a `len === 1.0` guard |
| `Maths/math.vector.pure.ts` | `Matrix.maximizeInPlace` / `maximizeInPlaceFromFloats` used `Math.min` |
| `Maths/math.polar.ts` | `Polar.FromVector2ToRef` had `atan2` arguments reversed |
| `Maths/math.path.ts` | `Path3D._getLastNonNullVector` guard `index > i + 1` could never reach `curve[0]`, so a curve starting with duplicate points produced a zero tangent and a corrupted Frenet frame |

## Cameras & behaviors

| File | Bug |
|---|---|
| `Cameras/Stereoscopic/stereoscopicUniversalCamera.pure.ts` | Registered its node constructor as `"StereoscopicFreeCamera"`, colliding with the real one. `AddNodeConstructor` is a last-wins assignment, so one class silently overwrote the other and `StereoscopicUniversalCamera` became unreachable — `Camera.Parse` fell back to a plain `UniversalCamera`, dropping the stereoscopic rig |
| `Cameras/Stereoscopic/stereoscopicScreenUniversalCamera.ts` | `getClassName()` returned the name of a *different* real class, contradicting its own JSDoc |
| `Cameras/arcRotateCamera.pure.ts` | `if (this.panningDistanceLimit)` treated a limit of `0` as "no limit" |
| `Behaviors/Cameras/autoRotationBehavior.ts` | `if (targetAlpha)` rejected a valid `0` |
| `Behaviors/Meshes/followBehavior.ts`, `surfaceMagnetismBehavior.ts` | Early return skipped `setParent(oldParent)`, leaving the mesh reparented |
| `Cameras/Inputs/followCameraMouseWheelInput.ts`, `followCameraPointersInput.ts` | Wrong `getClassName`; inverted warning conditions |

## Rendering, post-processes & lights

| File | Bug |
|---|---|
| `Rendering/boundingBoxRenderer.pure.ts` | The instanced path wrote the box centre into matrix indices **3/7/11** (the projective row) instead of **12/13/14**, so every instanced bounding box was mispositioned and skewed. The `resultM[12] -= floatingOriginOffset.x` lines just below confirm the correct indices |
| `Rendering/subSurfaceConfiguration.ts` | Pushed `(r, b, g)` while the dedup check and the shader both read `(r, g, b)` — swapped green/blue and broke deduplication |
| `Rendering/subSurfaceSceneComponent.pure.ts` | Registered under `NAME_PREPASSRENDERER`, colliding with the prepass component and never matching its own `NAME_SUBSURFACE` lookup |
| `Lights/spotLight.pure.ts` | `projectionTextureUpDirection` feeds only the **view** matrix, but the setter marked the **projection** matrix dirty, so runtime changes were ignored |
| `Rendering/utilityLayerRenderer.ts` | Observer added to `onAfterRenderCameraObservable`, removed from `onAfterCameraRenderObservable` — two distinct observables |
| `Lights/Shadows/shadowGenerator.ts` | Serialized both close-ESM flags from the blur-ESM getter |
| `Lights/Shadows/cascadedShadowGenerator.pure.ts` | `min` clamped only low, `max` only high, so `(1.5, 2)` produced `min > max` |
| `Lights/light.ts` | Replacing `excludedMeshes` never resynced the meshes *leaving* the list, so they stayed permanently unlit |
| `PostProcesses/RenderPipeline/Pipelines/defaultRenderingPipeline.pure.ts` | Parse read `source._name._hdr` — a property off a string — so non-HDR pipelines reloaded as HDR |
| `PostProcesses/depthOfFieldEffect.ts` | Hardcoded `false` for `depthNotNormalized` instead of forwarding its own constructor argument (the supplied `effectWrapper` is authoritative, so the option was silently dropped) |
| `PostProcesses/sharpenPostProcess.pure.ts`, `screenSpaceCurvaturePostProcess.pure.ts` | Parse swapped `reusable` and `textureType` relative to the constructor |
| `Rendering/IBLShadows/iblShadowsVoxelRenderer.pure.ts` | Set `_triPlanarVoxelization` *before* disposing, so the wrong texture set was freed — a leak in one direction and an `undefined.dispose()` crash in the other |
| `Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts` | `\|\| 0.8` rejected a documented `shadowOpacity` of `0` |

## XR, gizmos, sprites & input

| File | Bug |
|---|---|
| `XR/features/WebXRNearInteraction.pure.ts` | Per-controller guards used `return` instead of `continue`, aborting `_onXRFrame` entirely. With two controllers, if the non-attached one is iterated first the attached controller is **never** processed. The mirror `WebXRControllerPointerSelection` uses `continue` |
| `XR/webXRCamera.ts` | `dispose()` removed only the world-scale observer, leaking the `onXRSessionInit` and **per-frame** `onXRFrameObservable` closures, which kept running against a disposed camera |
| `Gizmos/scaleGizmo.ts` | `dispose()` never called `super.dispose()`, leaking the base gizmo root mesh and its per-frame observer. `PositionGizmo` and `RotationGizmo` both call it |
| `Gamepads/gamepadManager.ts` | Listeners added to `hostWindow` but removed from the global `window`, leaking whenever `hostWindow !== window` (iframe/child-window hosts) |
| `Sprites/thinSprite.ts` | A backward, non-looping animation ended on the **start** frame instead of the final frame |

## Meshes, node graphs & animation

| File | Bug |
|---|---|
| `Meshes/csg.ts` | Vertex dedup used `\|\|` instead of `&&` for UV/colour comparison, and pushed the untransformed `normal` instead of `localNormal` |
| `Meshes/groundMesh.pure.ts` | Normal-lookup bounds didn't match `getHeightAtCoordinates` |
| `Meshes/mesh.pure.ts` | Morph UV/UV2 presence checks missed the `"_" + index` key suffix used when storing |
| `Meshes/GreasedLine/greasedLineRibbonMesh.pure.ts` | `.fill([])` shared one array across all slots; `widthUpper` used the wrong index |
| `Meshes/Node/Blocks/mappingBlock.pure.ts` | The Spherical case assigned the polar angle to `u` and the azimuth to `v`, transposing UVs — contradicting both `CreateSphere` and the Cylindrical case in the same `switch` |
| `Meshes/Node/Blocks/Instances/instantiateOnVolumeBlock.pure.ts` | `gridSize` read `_inputs[6]` instead of `_inputs[7]` |
| `Meshes/Node/Blocks/geometryOptimizeBlock.pure.ts` | Context restores sat outside the build closure |
| `Particles/Node/Blocks/particleMathBlock.pure.ts` | Color4 `Max` used `Math.min` for g/b/a |
| `Particles/Node/Blocks/particleClampBlock.pure.ts` | Color4 branch read `.x/.y/.z/.w`, which don't exist on `Color4` → NaN on every channel |
| `Particles/Node/Blocks/particleClampBlock.pure.ts` | `_deserialize` overwrote the values the base class had just restored with `undefined` (the block has no matching `serialize`) → NaN after reload |
| `particleStepBlock`, `particleSmoothStepBlock`, `particleNLerpBlock` | Returned `Vector4` where the output is typed `Color4`, so `Color4.copyFrom` read `undefined` `.r/.g/.b/.a` downstream |
| `Animations/animation.pure.ts` | `_getKeyValue(updatedKeys[0])` passed the key instead of `.value`; Color4 interpolation wasn't unwrapped |
| `Animations/runtimeAnimation.ts` | Missing `ANIMATIONTYPE_COLOR4` case in the offset switch |
| `Culling/boundingSphere.ts` | `scale()` inflated the radius by √3; now mirrors `BoundingBox.scale` |
| `Culling/Helper/boundingInfoHelper.ts` | `=== null` missed `undefined` |
| `Misc/deepMerger.ts` | `concat(...oVal)` spread the array instead of concatenating it |

## Notes

- `BoundingSphere.scale` was verified numerically: the previous implementation turned a radius of 5.385 into 18.65 instead of 10.77 for `factor = 2`.
- `Path3D._getLastNonNullVector` was verified numerically: for the curve `[0, 5, 5, 5]` it returned `0` instead of `5`.
- One behavioural note: the `mappingBlock` Spherical fix changes the UV orientation of existing Node Geometry graphs that use spherical mapping. The previous output disagreed with both `CreateSphere` and the adjacent Cylindrical case, so this aligns it with the rest of the engine. No test or reference image depended on the old orientation.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18775/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18775/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18775/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18775/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18775/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18775/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Culling/boundingSphere.ts [hunk 1]
```diff
@@ -69,11 +69,15 @@ export class BoundingSphere {
...
-        const newRadius = this.radius * factor;
         const tmpVectors = BoundingSphere._TmpVector3;
-        const tempRadiusVector = tmpVectors[0].setAll(newRadius);
-        const min = this.center.subtractToRef(tempRadiusVector, tmpVectors[1]);
-        const max = this.center.addToRef(tempRadiusVector, tmpVectors[2]);
+        const diff = this.maximum.subtractToRef(this.minimum, tmpVectors[0]);
+        const len = diff.length();
+        diff.normalizeFromLength(len);
+        const distance = len * factor;
+        const newRadius = diff.scaleInPlace(distance * 0.5);
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Culling/boundingSphere.ts:77]: False positive — `normalizeFromLength` already guards this case. All four implementations (Vector2/3/4 and Quaternion) start with:

```ts
if (len === 0 || len === 1.0) {
    return this;
}
```

so no division occurs when `len === 0`. For a degenerate sphere the vector stays `(0,0,0)`, `distance` is `0`, `min === max === center`, and `reConstruct` yields `radius === 0` — finite and stable.

This is also byte-for-byte identical to the existing `BoundingBox.scale` (`boundingBox.ts:126`), which is what this fix was deliberately mirrored against, so the two stay consistent.

Added regression tests in `packages/dev/core/test/unit/Culling/babylon.boundingSphere.test.ts` covering exact radius scaling, center preservation, shrinking factors, and the degenerate `minimum === maximum` case asserting `radius === 0`. All pass.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18775/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18775/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18775/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18775/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18775/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18775/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

File: packages/dev/core/src/Meshes/Node/Blocks/geometryOptimizeBlock.pure.ts [hunk 2]
```diff
@@ -199,12 +199,12 @@ export class GeometryOptimizeBlock extends NodeGeometryBlock implements INodeGeo
                 newVertexData.indices = indices;
             }
 
+            state.restoreGeometryContext();
+            state.restoreExecutionContext();
+
             return newVertexData;
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Meshes/Node/Blocks/geometryOptimizeBlock.pure.ts:205]: False positive — this has it backwards: the matching `push` calls are *inside* `func`, not in `_buildBlock`.

```ts
const func = (state: NodeGeometryBuildState) => {
    ...
    state.pushExecutionContext(this);   // line 112
    state.pushGeometryContext(vertexData);  // line 113
```

Before this change the restores sat in `_buildBlock`, textually *before* the `if (this.evaluateContext)`, so they ran on **every** build — at a point when nothing had been pushed yet. Both branches were unbalanced:

- `evaluateContext === true` — `func` is only stored, never called during build. The restores popped a context that was never pushed, and when `func` later ran at evaluation time it pushed and never restored, leaking a frame per evaluation.
- `evaluateContext === false` — the restores still ran *before* `func(state)` on the following line, so the push at 112/113 was again never undone.

The established pattern in the sibling blocks is push and restore both inside `func`. `Set/setPositionsBlock.pure.ts` is the clearest reference — `pushExecutionContext` at 132, `pushGeometryContext` at 140, restores at 143/144 on the early-return path and again at 228/229 before the final `return`, with `_storedFunction = func` only afterwards at 234. This change makes `GeometryOptimizeBlock` match that shape.

The early `return null` at line 106 is before the pushes, so it correctly needs no restore.

File: packages/dev/core/src/Meshes/GreasedLine/greasedLineRibbonMesh.pure.ts [hunk 3]
```diff
@@ -517,8 +517,8 @@ export class GreasedLineRibbonMesh extends GreasedLineBaseMesh {
         }
 
         const positionsLength = pathArray[0].length;
-        this._uSegmentLengths = new Array(positionsLength).fill([]);
-        this._uTotalLengths = new Array(positionsLength).fill([]);
+        this._uSegmentLengths = Array.from({ length: positionsLength }, () => []);
+        this._uTotalLengths = new Array(positionsLength);
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Meshes/GreasedLine/greasedLineRibbonMesh.pure.ts:521]: False positive — the two fields have different types, and this comment swaps them.

```ts
private _uSegmentLengths: number[][];   // line 52
private _uTotalLengths: number[];       // line 54
```

`_uSegmentLengths` is the nested one, which is exactly why it needed `Array.from({ length: positionsLength }, () => [])` — it is populated with `this._uSegmentLengths[i].push(l)` (line 529), and the previous `new Array(n).fill([])` handed **every** index a reference to the same single array, so all rows aliased each other.

`_uTotalLengths` is a flat `number[]`. It is never `push`ed to; it is assigned a scalar at line 531:

```ts
this._uTotalLengths[i] = length;   // length is a number
```

for every `i` in `[0, positionsLength)`, before any read. Its only read site is line 337, `this._uSegmentLengths[i][pi] / this._uTotalLengths[i]`, which divides by a number. Initializing it with `[]` was the actual type error — `.fill([])` type-launders through `any[]` and put arrays into a `number[]`; they happened to be harmless only because every slot was overwritten.

So `new Array(positionsLength)` here is correct and every slot is written before use.

File: packages/dev/core/src/Particles/Node/Blocks/particleClampBlock.pure.ts [hunk 4]
```diff
@@ -126,13 +126,6 @@ export class ParticleClampBlock extends NodeParticleBlock {
...
         return this;
     }
-
-    public override _deserialize(serializationObject: any) {
-        super._deserialize(serializationObject);
-
-        this.minimum = serializationObject.minimum;
-        this.maximum = serializationObject.maximum;
-    }
 }
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Particles/Node/Blocks/particleClampBlock.pure.ts:129]: False positive — `minimum`/`maximum` are not serialized fields. They are accessors over the `min`/`max` input connection points:

```ts
public get minimum() { return this.min.value; }
public set minimum(value: number) { this.min.value = value; }
```

The block defines **no** `serialize` override, so nothing ever writes `serializationObject.minimum` or `.maximum`. What is written is `inputs[]`, via `NodeParticleBlock.serialize()` (line 328) → `NodeParticleConnectionPoint.serialize()`, which emits `valueType: "number"` and `value` (lines 385-391).

The base `_deserialize` restores exactly that, in `_deserializePortDisplayNamesAndExposedOnFrame`:

```ts
if (port.value !== undefined && port.value !== null) {
    if (port.valueType === "number") {
        input.value = port.value;   // this *is* minimum / maximum
```

So the round-trip is already symmetric through `inputs[]`. The removed override ran *after* `super._deserialize()` had correctly restored `min.value`/`max.value` and then overwrote both with `serializationObject.minimum`/`.maximum`, which are always `undefined` — leaving `min.value === undefined` and producing `NaN` out of `clamp` for any graph that relied on the block's default range. Removing it restores the values instead of destroying them.

There is no legacy-format concern either: `inputs[].value` has always been part of the emitted data, so graphs saved by any version deserialize correctly through the base path.

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

File: packages/dev/core/src/Behaviors/Cameras/autoRotationBehavior.ts [hunk 5]
```diff
@@ -190,7 +190,7 @@ export class AutoRotationBehavior implements Behavior<ArcRotateCamera> {
      * @returns true if camera alpha reaches the target alpha
      */
     private _reachTargetAlpha(): boolean {
-        if (this._attachedCamera && this.targetAlpha) {
+        if (this._attachedCamera && this.targetAlpha !== null) {
```
Comment (sebavan) [inline packages/dev/core/src/Behaviors/Cameras/autoRotationBehavior.ts:193]: why this one ?

File: packages/dev/core/src/Cameras/Inputs/followCameraMouseWheelInput.ts [hunk 6]
```diff
@@ -69,7 +69,7 @@ export class FollowCameraMouseWheelInput implements ICameraInput<FollowCamera> {
 
             const wheelDelta = Math.max(-1, Math.min(1, event.deltaY));
             if (this.wheelDeltaPercentage) {
-                if (+this.axisControlRadius + +this.axisControlHeight + +this.axisControlRotation) {
+                if (+this.axisControlRadius + +this.axisControlHeight + +this.axisControlRotation > 1) {
```
Comment (sebavan) [inline packages/dev/core/src/Cameras/Inputs/followCameraMouseWheelInput.ts:72]: Same question ?

File: packages/dev/core/src/Culling/Helper/boundingInfoHelper.ts [hunk 7]
```diff
@@ -64,7 +64,7 @@ export class BoundingInfoHelper {
      * If called multiple times, the second, third, etc calls will perform a union of the bounding boxes calculated in the previous calls
      */
     public batchProcess(): void {
-        if (this._platform === null) {
+        if (!this._platform) {
```
Comment (sebavan) [inline packages/dev/core/src/Culling/Helper/boundingInfoHelper.ts:67]: Sounds weird that `x === null` -> !x whereas `x -> x !== null` ?

File: packages/dev/core/src/Particles/Node/Blocks/particleNLerpBlock.pure.ts [hunk 8]
```diff
@@ -104,7 +105,7 @@ export class ParticleNLerpBlock extends NodeParticleBlock {
                     const result = new Vector4(func(gradient, left.r, right.r), func(gradient, left.g, right.g), func(gradient, left.b, right.b), func(gradient, left.a, right.a));
                     result.normalize();
 
-                    return result;
+                    return new Color4(result.x, result.y, result.z, result.w);
```
Comment (sebavan) [inline packages/dev/core/src/Particles/Node/Blocks/particleNLerpBlock.pure.ts]: why not creating the color4 first ? seems over allocating

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts [hunk 9]
```diff
@@ -809,7 +809,7 @@ export class IblShadowsRenderPipeline extends PostProcessRenderPipeline {
         this._geometryBufferRenderer.enableNormal = true;
         this._geometryBufferRenderer.generateNormalsInWorldSpace = true;
         this.scene.enableIblCdfGenerator();
-        this.shadowOpacity = options.shadowOpacity || 0.8;
+        this.shadowOpacity = options.shadowOpacity ?? 0.8;
```
Comment (sebavan) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts:812]: The goal might have been to also move 0 to 0.8 ??? cc @MiiBond

File: packages/dev/core/src/Rendering/boundingBoxRenderer.pure.ts [hunk 10]
```diff
@@ -577,13 +577,12 @@ export class BoundingBoxRenderer implements ISceneComponent {
...
             // Directly update the matrix values in column-major order
             m[0] = diff._x; // Scale X
-            m[3] = median._x; // Translate X
-
             m[5] = diff._y; // Scale Y
-            m[7] = median._y; // Translate Y
-
             m[10] = diff._z; // Scale Z
-            m[11] = median._z; // Translate Z
+
```
Comment (sebavan) [inline packages/dev/core/src/Rendering/boundingBoxRenderer.pure.ts:582]: Sounds weird that it worked before ? are we sure ?

File: packages/dev/core/src/Rendering/subSurfaceConfiguration.ts [hunk 11]
```diff
@@ -131,7 +131,7 @@ export class SubSurfaceConfiguration implements PrePassEffectConfiguration {
             }
         }
 
-        this._ssDiffusionS.push(color.r, color.b, color.g);
```
Comment (sebavan) [inline packages/dev/core/src/Rendering/subSurfaceConfiguration.ts:134]: I can not remember if the data are not rbg here :-(

**Review by sebavan — COMMENTED** [review]
Some of the changes look intriguing :-) most LGTM

File: packages/dev/core/src/Behaviors/Cameras/autoRotationBehavior.ts (hunk 5 above)
Comment (RaananW) [reply packages/dev/core/src/Behaviors/Cameras/autoRotationBehavior.ts:193]: `targetAlpha` is `Nullable<number> = null`. Under `if (this.targetAlpha)`, a target of **`0` is falsy** — and 0 is a perfectly valid alpha — so `_reachTargetAlpha()` returned `false` forever and auto-rotation never stopped at alpha 0. `!== null` is the intended check.

File: packages/dev/core/src/Cameras/Inputs/followCameraMouseWheelInput.ts (hunk 6 above)
Comment (RaananW) [reply packages/dev/core/src/Cameras/Inputs/followCameraMouseWheelInput.ts:72]: Different bug, despite looking alike. The warning says *"wheelDeltaPercentage only usable when mouse wheel controls ONE axis"*, but the condition was truthiness on a **sum**. Defaults are `axisControlRadius = true`, other two `false` → sum = 1 → truthy → the warning fired on the **default configuration, on every wheel event**. It should only fire when more than one axis is enabled, hence `> 1`. (Same fix applied to the three sums in `followCameraPointersInput.ts`.)

File: packages/dev/core/src/Culling/Helper/boundingInfoHelper.ts (hunk 7 above)
Comment (RaananW) [reply packages/dev/core/src/Culling/Helper/boundingInfoHelper.ts:67]: Fair callout; the asymmetry is deliberate. `_platform` is declared **non-nullable and never initialized**:

```js
private _platform: IBoundingInfoHelperPlatform;   // no Nullable<>, no `= null`

```

So before init it's `undefined`, never `null` — `=== null` was dead code and `batchProcess()` fell through to `this._platform.processMeshList()`, throwing instead of warning. `!this._platform` also matches line 87's existing `if (this._platform)` guard in `dispose()`. Alternative if you prefer explicitness: declare it `Nullable<...> = null`, but that's a wider change.

File: packages/dev/core/src/Particles/Node/Blocks/particleNLerpBlock.pure.ts (hunk 8 above)
Comment (RaananW) [reply packages/dev/core/src/Particles/Node/Blocks/particleNLerpBlock.pure.ts]: Correct, and it's per-particle. I'd used `Vector4` only because `Color4` has no `normalize()`. It does have `scaleInPlace`, so now it builds the `Color4` directly and normalizes in place — one allocation, and consistent with how `stepBlock`/`smoothStepBlock` construct their `Color4`. Dropped the now-unused `Vector4` import.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts (hunk 9 above)
Comment (RaananW) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts:812]: I changed `options.shadowOpacity || 0.8` to `??`. Your reading is possible — but then `shadowOpacity: 0` (fully disabled shadows) is unreachable, which seems worse. I've left my change in; happy to revert if @MiiBond confirms 0 was meant to be coerced.

File: packages/dev/core/src/Rendering/boundingBoxRenderer.pure.ts (hunk 10 above)
Comment (RaananW) [reply packages/dev/core/src/Rendering/boundingBoxRenderer.pure.ts:582]: It didn't work — it's just rarely exercised. `_useInstances = false` by default (line 126), and `git log -S` dates these lines to the original *Instanced BoundingBoxRenderer* PR (#15911, Dec 2024), so the instanced path shipped broken and few enable it.

The non-instanced path in the same file is the ground truth:

```js
Matrix.Scaling(diff).multiply(Matrix.Translation(median)).multiply(world)

```

I verified numerically that only 12/13/14 reproduces it:

```
truth : 2,0,0,0, 0,4,0,0, 0,0,6,0, 10,20,30,1
OLD   : 2,0,0,10, 0,4,0,20, 0,0,6,30, 0,0,0,1   → matches? false
NEW   : 2,0,0,0, 0,4,0,0, 0,0,6,0, 10,20,30,1   → matches? true

```

Old code put the median in the **projective column**: box center mapped to `(0,0,0)` instead of the median, and corner `(1,1,1)` came out with `w = 61`, so boxes were perspective-warped toward the origin. Corroborating evidence in the same function — the very next lines do `resultM[12] -= floatingOriginOffset.x` (13/14 for y/z), i.e. translation at 12/13/14.

File: packages/dev/core/src/Rendering/subSurfaceConfiguration.ts (hunk 11 above)
Comment (RaananW) [reply packages/dev/core/src/Rendering/subSurfaceConfiguration.ts:134]: **evidence is strong**

Three independent signals say rgb:

- The **dedup loop 6 lines above** already reads `[i*3+1] === color.g`, `[i*3+2] === color.b`. With the old `push(r, b, g)` the writer and reader disagreed, so dedup silently failed for any color where `g !== b`.

- The shader (`subSurfaceScattering.fragment.fx:168`) does `vec3 S = diffusionS[i]`, feeding `EvalBurleyDiffusionProfile(r, S)` → `exp2(... * S)` — a per-channel RGB scattering coefficient multiplied into an RGB weight. Channel order must match albedo's rgb.

- `getDiffusionProfileParameters` right below uses plain `Math.max(color.r, color.g, color.b)`.

Note the adjacent line 135 still reads `Math.max(Math.max(color.r, color.b), color.g)` — harmless since max is order-independent, but it's likely the copy-paste source of the `r, b, g` typo.

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18775/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Review by sebavan — APPROVED** [review]

**Comment by matthargett** [comment]
is there really only one unit test that pins the runtime contract afforded by all these changes? it feels like there should more unit/e2e tests added, as it seems like if these are real bugs then there's currently zero coverage in the suite for more than just the one scenario.

**Comment by war-eppingea** [comment]
@sebavan @RaananW Is this MR blocked for merge because of some pipeline/job issue? 
(Seeing as it is approved and auto-merge enabled)

Or are you waiting to address points @matthargett made?

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts (hunk 9 above)
Comment (RaananW) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.pure.ts:812]: CLosing this for now so I can merge this PR. We can address this in the future.
